examples/opensora_hpcai/opensora/datasets/video_dataset_refactored.py

```diff
@@ -219,27 +219,36 @@ def _get_item(self, idx: int) -> Tuple[Any, ...]:
             data["video"] = (vae_latent * self._vae_scale_factor).astype(np.float32)
 
         else:
-            with VideoReader(data["video"]) as reader:
-                min_length = self._min_length
-                if self._buckets:
-                    data["bucket_id"] = self._buckets.get_bucket_id(
-                        T=len(reader), H=reader.shape[1], W=reader.shape[0], frame_interval=self._stride
+            reader = VideoReader(data["video"])
+            min_length = self._min_length
+            if self._buckets:
+                if "shape" not in data:
+                    # decord doesn't support reading the video shape until a frame is fetched
+                    # so read the shape with OpenCV and cache it for further iterations
+                    cap = cv2.VideoCapture(data["video"], apiPreference=cv2.CAP_FFMPEG)
```

Comment:
can read shape from reader. 
```
frame_h, frame_w, _ = reader[0].shape
```

Reply:
I didn't have time to benchmark, but I assume that reading video metadata is much cheaper than reading and decoding the first video frame (especially if it's 1080p+ with a high bitrate).